### DISCUSSION TITLE
About replacement tokens

### DISCUSSION THREAD

**Discussion by SergKorol** [Tyrrrz/GitHubActionsTestLogger#23] [Q&A]
I found the replacement tokens aren't working. Does it a bug, or I something do wrong?

**Comment by Tyrrrz** [comment]
What are you trying to do?

**Reply by SergKorol** [reply]
I put a command from the tutorial with custom annotations. The ordinary strings work fine however, variables like $test, $trace is not.
For instance: --logger "GitHubActions;annotations.titleFormat=$test;annotations.messageFormat=$error"

**Reply by Tyrrrz** [reply]
What happens? Do they not get replaced? Do they get replaced by blank strings? Do you have a link to the workflow?

**Reply by SergKorol** [reply]
Yes, blank strings and logging doesn't work when the strings are empty. I'm getting something like GitHubActions;annotations.titleFormat=;annotations.messageFormat=
But if I'll make GitHubActions;annotations.titleFormat=MyTitle;annotations.messageFormat=MyError
Then it works fine.
Example of my job:
- name: Test
      working-directory: src/ShopManagementSystem
      run: > 
        dotnet test 
        --no-build 
        --verbosity normal 
        --logger "GitHubActions;annotations.titleFormat=$test;annotations.messageFormat=$error"

**Reply by Tyrrrz** [reply]
Can't reproduce, works for me

![image](https://github.com/Tyrrrz/GitHubActionsTestLogger/assets/1935960/e451f80f-51c4-485f-afec-7333f16c0d30)

**Reply by SergKorol** [reply]
I just now grabbed your demo project and locally ran it. I managed to reproduce this issue in your demo.
<img width="1337" alt="image" src="https://github.com/Tyrrrz/GitHubActionsTestLogger/assets/24244973/5f1158a3-fc8a-4ed2-bf6f-83889cce55e8">
<img width="1147" alt="image" src="https://github.com/Tyrrrz/GitHubActionsTestLogger/assets/24244973/b3f524a3-5e9a-47ce-94bf-46e4b0e34be8">

**Reply by Tyrrrz** [reply]
Hmm, that's really weird. Could it be that `$abc` values are replaced by your terminal or something?

**Reply by SergKorol** [reply]
I checked it in Docker — the same issue.

**Reply by Tyrrrz** [reply]
I checked in Docker and was able to reproduce this. It seems that `$` is treated as a variable indicator on Linux/macOS shells.

Try escaping it as `\$`, or use single quotes instead of double quotes: 

```sh
dotnet test -p:IsTestProject=true -l 'GitHubActions;annotations.titleFormat=Foo_$test_Bar'
```

**Reply by SergKorol** [reply]
Yes, with \\$test works fine. I would appreciate it if you changed the documentation dedicated to this issue. Thanks a lot for your time.

**Reply by Tyrrrz** [reply]
Looking at this issue, it seems that there's no way to uniformly handle `$` on all platforms. On Windows it doesn't need escaping and trying to escape it messes up the formatting. Maybe I should try to find a different character to use as a delimiter. Thanks for reporting this.

**Reply by Tyrrrz** [reply]
I've switched to using `@`: https://github.com/Tyrrrz/GitHubActionsTestLogger/blob/master/Changelog.md#v232-01-jun-2023

**Reply by Tyrrrz** [reply]
By the way, what custom format are you using inside your workflows? I'm curious
